applications/StatisticsApplication/custom_methods/temporal_variance_method.h

```diff
@@ -0,0 +1,310 @@
+            Params["output_mean_variables"].GetStringArray();
+        const std::vector<std::string>& output_variable_2_names_list =
+            Params["output_variance_variables"].GetStringArray();
+
+        std::vector<TemporalMethod::Pointer> method_list;
+        if (rNormType == "none") // for non norm types
+        {
+            MethodUtilities::CheckInputOutputVariables(
+                input_variable_names_list, output_variable_1_names_list);
+            MethodUtilities::CheckInputOutputVariables(
+                input_variable_names_list, output_variable_2_names_list);
+
+            const int number_of_variables = input_variable_names_list.size();
+            for (int i = 0; i < number_of_variables; ++i)
+            {
+                const std::string& r_variable_input_name = input_variable_names_list[i];
+                const std::string& r_variable_1_output_name =
+                    output_variable_1_names_list[i];
+                const std::string& r_variable_2_output_name =
+                    output_variable_2_names_list[i];
+                ADD_TEMPORAL_VALUE_METHOD_TWO_OUTPUT_VARIABLE_OBJECT(
+                    rModelPart, rNormType, r_variable_input_name, EchoLevel,
+                    r_variable_1_output_name, r_variable_2_output_name, method_list, ValueMethod)
+            }
+        }
+        else // for values with norms
+        {
+            MethodUtilities::CheckVariableType<double>(output_variable_1_names_list);
+            MethodUtilities::CheckVariableType<double>(output_variable_2_names_list);
+
+            const int number_of_variables = input_variable_names_list.size();
+            for (int i = 0; i < number_of_variables; ++i)
+            {
+                const std::string& r_variable_input_name = input_variable_names_list[i];
+                const std::string& r_variable_1_output_name =
+                    output_variable_1_names_list[i];
+                const std::string& r_variable_2_output_name =
+                    output_variable_2_names_list[i];
+                ADD_TEMPORAL_NORM_METHOD_TWO_OUTPUT_VARIABLE_OBJECT(
+                    rModelPart, rNormType, r_variable_input_name, EchoLevel,
+                    r_variable_1_output_name, r_variable_2_output_name, method_list, NormMethod)
+            }
+        }
+
+        return method_list;
+
+        KRATOS_CATCH("");
+    }
+
+private:
+    template <typename TDataType>
+    void static CalculateMeanAndVariance(TDataType& rMean,
+                                         TDataType& rVariance,
+                                         const TDataType& rNewDataPoint,
+                                         const double DeltaTime,
+                                         const double TotalTime)
+    {
+        const double new_total_time = TotalTime + DeltaTime;
+        const TDataType new_mean =
+            (rMean * TotalTime + rNewDataPoint * DeltaTime) * (1.0 / new_total_time);
+        rVariance =
```

Comment:
Hi,

So I did the basis of this in my MSc Thesis. We discussed with Suneth and re-derived all equations for "on the fly" calculation. There is no specific reference. Take the mean for example:

1. you have the sample arithmetic mean for n values
2. than you write the sample arithmetic mean for n+1 values
3. you formulate a recursive procedure for n+1 values which takes the results for the previous n
4. you generalize this

That is it. In mathematics/statistics books you will either find the definition for infinite number of values or n finite ones, depending on the function. Steps 2 and 3 are just a reordering of basic mathematic expressions like +,-, *,/. In this sense I think your comment is answered. Please note this explanation for the next occurrences as well. Also remove your blocking of the merge, if possible, as the definition and implementation is per se not wrong. The lack of reference is because it is trivial in the sense of deriving it as in the steps above, also in the readme.

As long as the implementation is documented, it is up to the:

- user to check if the definition corresponds to what she/he wants to achieve
- other developers to add new functionalities

Reply:
I agree for the mean. My doubt is for the variance, which is a non linear operator, thus does not work as the mean.

Please have a look at the test I add below (I cannot add it here, no clue why). The test is what is implemented for constant time steps.

For a time window of `10` realizations (`time_step*len(values)`), the difference I get between numpy variance and the variance formula of the README (line 152) is of the order `53822`.
For a time window of 100 realizations, I get a difference of `15802`. Of course it goes to zero as the number of realizations go to infinity.

Please, let me know if something is wrong in my test.

If it may help you, I have worked with strategies as the ones described in [eq. 2.12](https://link.springer.com/article/10.1007/s00180-015-0637-z).

Reply:
Existing implementation works well with higher number of realizations as well. The formulation is little bit different, and I made it compatible with dynamic time stepping as well. (Please check the attachement. I also cannot attach files to this comment)

Reply:
Ok then, now results look good. I think the README has my implementation, not this one. Am I wrong?

Reply:
I only included basic definitions, not the exact definition, It is hard to write down exact definition in readme file. (lot of mathematical lines need to be there in latex, and I'm not sure README is the place to have extensive developement of formulations).

Reply:
Sorry for misguiding via README, I only added README to get an idea about how to use capabilities in StatisticsApplication, and give basic definitions. Not the detailed ones

Reply:
Up to you, since you are the developer, but people may think the same I thought if they see that formula. In any case, I resolve the conversation!

Reply:
No problem :)